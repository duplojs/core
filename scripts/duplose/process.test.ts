import { Process } from "./process";
import { BuildNoRegisteredDuploseError, CutStep, Duplose, Hook, zod } from "..";
import { Request } from "@scripts/request";
import { PreflightStep } from "@scripts/step/preflight";
import { Response } from "@scripts/response";
import { CheckpointList } from "@test/utils/checkpointList";
import { DuploTest } from "@test/utils/duploTest";
import { createProcessDefinition } from "@test/utils/manualDuplose";
import { ExtractStep } from "@scripts/step/extract";
import { getTypedEntries } from "@duplojs/utils";

describe("Process", () => {
	const checkpointList = new CheckpointList();
	const duplo = new DuploTest({ environment: "TEST" });

	const preflightProcess = new Process(createProcessDefinition({}));
	preflightProcess.instance = duplo;
	const preflight = new PreflightStep(preflightProcess, { pickup: ["flute"] as any });
	const cutStep = new CutStep(
		({ dropper }) => {
			checkpointList.addPoint("cut");
			return dropper({ toto: "true" });
		},
		["toto"],
		[new Response(100, "toto", zod.undefined())],
	);
	const extractStep = new ExtractStep(
		{
			params: {
				userId: zod.coerce.number(),
			},
		},
	);

	const process = new Process(createProcessDefinition({
		steps: [cutStep, extractStep],
		preflightSteps: [preflight],
		drop: ["toto", "userId"],
		name: "test",
	}));

	it("name", () => {
		expect(process.definiton.name).toBe("test");
	});

	it("getAllHooks", () => {
		const hooks = process.getAllHooks();

		getTypedEntries(hooks)
			.forEach(([key, value]) => {
				if (!(value instanceof Hook)) {
					return;
				}

				expect(value.subscribers[0])
					.toBe(process.hooks[key]);

				expect((value.subscribers[1] as Hook).subscribers[0])
					.toBe(preflightProcess.hooks[key]);
			});
	});

	it("hasDuplose", () => {
		expect(process.hasDuplose(new Process(createProcessDefinition()))).toBe(false);
		expect(process.hasDuplose(process)).toBe(true);
		expect(process.hasDuplose(preflightProcess)).toBe(true);
	});

	it("build", async() => {
		checkpointList.reset();
		const spy = vi.spyOn(Duplose.defaultEvaler, "makeFunction");

		await expect(() => process.build()).rejects.toThrowError(BuildNoRegisteredDuploseError);

		process.instance = duplo;

		await process.build();

		expect(spy).toBeCalled();

		expect(spy.mock.lastCall?.[0].content).toMatchSnapshot();

		const processFunction = await process.build();

		expect(processFunction(new Request({ params: { userId: "2" } } as any), undefined, undefined)).toStrictEqual({
			toto: "true",
			userId: 2,
		});

		expect(checkpointList.getPointList()).toStrictEqual([
			"start",
			"cut",
			"end",
		]);
	});
});
