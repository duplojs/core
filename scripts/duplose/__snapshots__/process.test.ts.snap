// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Process > build 1`] = `
"
		let floor = this.makeFloor();
		let result = undefined;
		floor.drop("options", options);
		floor.drop("input", input);
		Execution: {
			
/* preflight-before */
/* end_block */


			
				
/* preflight-process-(0)-before */
/* end_block */


				result = this.preflightSteps[0].processFunction(
					request,
					this.preflightSteps[0].params.options,
					this.preflightSteps[0].params.input(floor.pickup)
				);

				
/* preflight-process-(0)-before-check-result */
/* end_block */


				if(result instanceof this.Response){

break Execution;
}

				
/* preflight-process-(0)-before-drop */
/* end_block */


				floor.drop("flute", result["flute"]);

				
/* preflight-process-(0)-after */
/* end_block */

			

			
/* preflight-after */
/* end_block */


			
/* steps-before */
/* end_block */


			
		
/* step-cut-(0)-before */
/* end_block */


		result = this.steps[0].cutFunction(floor, request);

		
/* step-cut-(0)-before-check-result */
/* end_block */


		if(result instanceof this.Response){

				let temp = this.steps[0].responseZodSchema.safeParse(result);

				if(!temp.success){
					throw new this.ContractResponseError(temp.error, result);
				}
			
break Execution;
}

		
/* step-cut-(0)-before-drop */
/* end_block */


		floor.drop("toto", result["toto"]);

		
/* step-cut-(0)-after */
/* end_block */

		

/* setp-extract-(1)-before */
/* end_block */


			
/* setp-extract-(1)-(params)-(userId)-before */
/* end_block */

		
			{
				let temp = this.steps[1]
					.extractObject["params"]["userId"]
					.safeParse(request["params"]?.["userId"])
		
				if(!temp.success){
					result = this.steps[1].catchError(
						"params",
						"userId",
						temp.error,
					);
					break Execution;
				}
		
				floor.drop(
					"userId",
					temp.data,
				);
			}
		
			
/* setp-extract-(1)-(params)-(userId)-after */
/* end_block */

			

/* setp-extract-(1)-after */
/* end_block */


			
/* steps-after */
/* end_block */

		}

		
/* process-if-return-before */
/* end_block */

		if(result instanceof this.Response){
			
/* process-result-return-before */
/* end_block */

			return result;
		}
		else {
			
/* process-drop-return-before */
/* end_block */

			return {
				"toto": floor.pickup("toto"),
"userId": floor.pickup("userId"),
			};
		}
		"
`;
